src/main/src/main/java/org/geoserver/security/filter/GeoServerSecurityContextPersistenceFilter.java

```diff
@@ -36,31 +39,51 @@ public void initializeFromConfig(SecurityNamedServiceConfig config) throws IOExc
                 (SecurityContextPersistenceFilterConfig) config;
 
         HttpSessionSecurityContextRepository repo = new HttpSessionSecurityContextRepository();
-        SecurityContextPersistenceFilter filter =
-                new SecurityContextPersistenceFilter(repo) {
+        OncePerRequestFilter eagerFilter =
+                new OncePerRequestFilter() {
                     @Override
-                    public void doFilter(ServletRequest req, ServletResponse res, FilterChain chain)
+                    public void doFilterInternal(
+                            HttpServletRequest request,
+                            HttpServletResponse response,
+                            FilterChain chain)
                             throws IOException, ServletException {
                         // set the hint for authentcation servlets
-                        req.setAttribute(ALLOWSESSIONCREATION_ATTR, isAllowSessionCreation);
-                        if (isAllowSessionCreation)
-                            ((HttpServletRequest) req).getSession(); // create session if allowed
+                        request.setAttribute(ALLOWSESSIONCREATION_ATTR, isAllowSessionCreation);
+                        if (isAllowSessionCreation) {
+                            request.getSession();
+                        }
                         // set the hint for other components
-                        req.setAttribute(
+                        request.setAttribute(
                                 GeoServerSecurityFilterChainProxy.SECURITY_ENABLED_ATTRIBUTE,
                                 Boolean.TRUE);
-                        super.doFilter(req, res, chain);
+                        // ensure that filter is only applied once per request
+                        if (request.getAttribute(FILTER_APPLIED) != null) {
+                            chain.doFilter(request, response);
+                            return;
+                        }
```

Comment:
Correct.  It appears to load the request into the context repository if this is the first time this filter is called.  Otherwise it just ensures that the session has been created and then calls chain.doFilter before returning.  Should this approach change in the OncePerRequestFilter implementation?

Reply:
No need to change, I was just wondering. Others might too, maybe it's a good idea to link to the old code for reference.

Reply:
link added in note